src/sinks/aws_s3/sink.rs

```diff
@@ -1,28 +1,192 @@
+            framer.as_ref().clone(),
+            serializer.clone(),
+        ));
+
+        let builder_limit = NonZeroUsize::new(64);
+
+        input
+            .map(|event| {
+                let mut to_encode = event.clone();
+                transformer.transform(&mut to_encode);
+
+                let mut encoded = BytesMut::new();
+                serializer.encode(to_encode, &mut encoded).unwrap();
+
+                EncodedEvent {
+                    inner: event,
+                    encoded,
+                }
+            })
+            .batched_partitioned(partitioner, batcher_settings)
+            .filter_map(|(key, batch)| async move { key.map(move |k| (k, batch)) })
+            .concurrent_map(builder_limit, move |(partition_key, encoded_events)| {
+                let framer = Arc::clone(&framer);
+                let combined_encoder = Arc::clone(&combined_encoder);
+                let options = Arc::clone(&options);
+
+                Box::pin(async move {
+                    // This is silly because we really just need the prefix, delimiter, and suffix. Oh well.
+                    let mut framer = framer.as_ref().clone();
+
+                    let mut grouped_sizes = GroupedCountByteSize::new_tagged();
+                    let mut events = Vec::with_capacity(encoded_events.len());
+                    let mut encoded = Vec::with_capacity(encoded_events.len());
+                    for e in encoded_events {
+                        grouped_sizes.add_event(&e.inner, e.encoded.len().into());
+                        events.push(e.inner);
+                        encoded.push(e.encoded);
+                    }
+
+                    // TODO: this doesn't include framing, is that right?
+                    let events_encoded_size = encoded.iter().map(BytesMut::len).sum::<usize>();
+
+                    let finalizers = events.take_finalizers();
+                    let s3_key_prefix = partition_key.key_prefix.clone();
+
+                    let metadata = S3Metadata {
+                        partition_key,
+                        s3_key: s3_key_prefix,
+                        finalizers,
+                    };
+
+                    // TODO: not doing compression yet
+                    let mut payload = BytesMut::new();
+                    payload.extend_from_slice(combined_encoder.batch_prefix());
+                    let mut remaining = encoded.len();
+                    for buf in encoded {
+                        payload.extend_from_slice(buf.as_ref());
+                        remaining -= 1;
+                        if remaining > 0 {
+                            // write the frame delimiter
+                            framer.encode((), &mut payload).expect("framing to bytes");
+                        }
+                    }
+                    payload.extend_from_slice(combined_encoder.batch_suffix());
+
+                    let request_metadata = RequestMetadata::new(
```

Comment:
Seems like a fringe benefit of this would be that we don't need the `RequestMetadataBuilder` anymore ? I added that last year and it was kind of annoying, like if we have all the information we need in one place there are definite perks to that.